QuizApp/MVP/Presenter/LoginPresenter.swift

```diff
@@ -0,0 +1,84 @@
+//
+//  LoginPresenter.swift
+//  QuizApp
+//
+//  Created by Danijel Stracenski on 15.05.2021..
+//
+
+import Foundation
+import Reachability
+
+struct LoginResponse:Codable{
+    let token:String
+    let userId:Int
+    
+    enum CodingKeys: String, CodingKey {
+        case token
+        case userId = "user_id"
+    }
+}
+
+protocol LoginViewDelegate: AnyObject {
+    func showGoodLogin()
+    func showLoginError(error:RequestError)
+    func showReachabilityError()
+    // bad input, no need for alert, only mark input fileds
+    func showLoginClientError()
+
+    func hidePassword(existingText:String)
+    func showPassword()
+}
+
+class LoginPresenter{
+    weak var delegate:LoginViewDelegate!
+    var networkService:NetworkServiceProtocol!
+    
+    
+    init(delegate:LoginViewDelegate){
+        self.delegate = delegate
+        self.networkService = NetworkService()
+    }
+    
+    func login(username: String, password: String) {
+        DispatchQueue.global(qos: .userInitiated).async {
+            if !self.networkService.checkReachability(){
+                self.delegate.showReachabilityError()
+                return
+            }
+            
+            let url = URL(string: "https://iosquiz.herokuapp.com/api/session?username=\(username)&password=\(password)")!
+            var request = URLRequest(url: url)
+            request.httpMethod = "POST"
+            
+            self.networkService.executeUrlRequest(request) { (result: Result<LoginResponse, RequestError>) in
+                switch result {
+                    case .failure(let error):
+                        switch error {
+                        case .clientError:
+                            self.delegate.showLoginClientError()
+                        default:
+                            self.delegate.showLoginError(error:error)
+                        }
+                    case .success(let value):
+                        let defaults = UserDefaults.standard
+                        defaults.set(value.userId, forKey: "user_id")
+                        defaults.set(value.token, forKey: "user_token")
+                        print("username:\(username), password:\(password),\nuserId:\(value.userId), token:\(value.token)")
+                        self.delegate.showGoodLogin()
+                }
+            }
+        }
+    }
```

Comment:
This is ok, better practice is to take this whole logic and put it inside NetworkService and expose it with protocol.

To make it clearer:

This is your protocol:
```swift 
protocol NetworkServiceProtocol{
    func checkReachability()->Bool
    func executeUrlRequest<T: Decodable>(_ request: URLRequest, completionHandler:
    @escaping (Result<T, RequestError>) -> Void)
}
```

Think about network service as a tool that gives you results of network requests. In your homework you don't have a lot of requests so that can all be done in the same class (NetworkService).
Take this functions ->
```swift
func checkReachability()->Bool
func executeUrlRequest<T: Decodable>(_ request: URLRequest, completionHandler:
@escaping (Result<T, RequestError>) -> Void)
```

You can make them private inside NetworkService and use them for triggering login, fetchingQuizzes and so on.
If you design it that way, your NetworkServiceProtocol will expose only concrete tasks for your presenters.
With that approach your NetworkServiceProtocol should look something like this ->

```swift 
protocol NetworkServiceProtocol {
     func login(username: String, password: String, callback: @escaping (LoginResponse) -> Void)
}
```

Of course you need all other functions inside that protocol like fetchQuizzes, sendQuizResult and fetchLeaderboard.

Don't take this the wrong way, your solution is good but I can see that you have put a lot of effort in this homework, so I just wanted to show you the best practices for doing this.

Reply:
Will do, this makes more sense then for all the logic to be in presenter. Thanks for the feedback.